File: valid-word-abbreviation/valid-word-abbreviation.js
/**
 * @param {string} word
 * @param {string} abbr
 * @return {boolean}
 */
var validWordAbbreviation = function(word, abbr) {

    let j = 0;
    
    let skip = 0;
    for(let i=0; i < abbr.length; i++){
        
         if(!isNaN(abbr[i])){
              let num = Number(abbr[i]);
              if(!skip && !num) return false; //if there is a leading zero return false
             skip = skip * 10 + Number(abbr[i])
         }else{
             j += skip
            if(word[j] != abbr[i])return false
            skip = 0  
            j++;
         }
    }
    return j + skip == word.length
};

File: valid-word-abbreviation/valid-word-abbreviation.py
class Solution:
    def validWordAbbreviation(self, word: str, abbr: str) -> bool:
        
        i , j = 0,0
        
        while i < len(word) and j < len(abbr):
            
            if abbr[j].isdigit() and abbr[j] > '0':
                num = abbr[j]
                
                while j+1< len(abbr) and abbr[j+1].isdigit():
                    j +=1
                    num +=abbr[j]
                
                i += int(num)-1
            else:
                
                if abbr[j] != word[i]:
                    return False
            i ,j = i +1, j +1
        
        
        print(i,j)
        
        return  i==len(word) and j==len(abbr)
        